wss/wss.go

```diff
@@ -0,0 +1,186 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wss
+
+import (
+	"bufio"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+
+	info "github.com/google/cadvisor/info/v1"
+	"github.com/google/cadvisor/stats"
+	"k8s.io/klog"
+)
+
+const (
+	cgroupProcfs = "cgroup.procs"
+)
+
+var (
+	smapsFilePathPattern     = "/proc/%s/smaps"
+	clearRefsFilePathPattern = "/proc/%s/clear_refs"
+
+	referencedRegexp = regexp.MustCompile(`Referenced:\s*([0-9]+) kB`)
+	isDigitRegExp    = regexp.MustCompile("\\d+")
+)
+
+type wssManager struct {
+	resetInterval uint64
+	wssEnabled    bool
+}
+
+func NewManager(resetInterval uint64, wssEnabled bool) stats.Manager {
+	if resetInterval == uint64(0) {
+		klog.Warningf("Incorrect value of wss_reset_interval, currently set to %d, working set size metric cannot be provided", resetInterval)
+		wssEnabled = false
+	}
+	return &wssManager{resetInterval: resetInterval, wssEnabled: wssEnabled}
+}
+
+func (m *wssManager) GetCollector(cgroupPath string) (stats.Collector, error) {
+	if !m.wssEnabled {
+		return &stats.NoopCollector{}, nil
+	}
+	collector := newCollector(cgroupPath, m.resetInterval)
+	return collector, nil
+}
+
+func (m *wssManager) Destroy() {}
+
+func newCollector(cgroupPath string, resetInterval uint64) *wssCollector {
+	collector := &wssCollector{cgroupCPUPath: cgroupPath, resetInterval: resetInterval}
+	return collector
+}
+
+// wssCollector holds information necessary to calculate working set size
+type wssCollector struct {
+	// cgroupCPUPath groupCPUPath CPU cgroup path
+	cgroupCPUPath string
+	// Cycles counter for measurements cycles
+	cycles uint64
+	// resetInterval number of measurement csycles after which referenced bytes are cleared
+	resetInterval uint64
+}
+
+// GetStat calculate working set size and clear referenced bytes
+// see:  https://github.com/brendangregg/wss#wsspl-referenced-page-flag
+func (w *wssCollector) UpdateStats(stats *info.ContainerStats) error {
+	w.cycles++
+
+	pids, err := w.getPids()
+	if err != nil {
+		return err
+	}
+
+	referencedKBytes, err := w.getReferenced(pids)
+	if err != nil {
+		return err
+	}
+
+	err = w.clearReferenced(pids)
+	if err != nil {
+		return err
+	}
+
+	stats.Wss = referencedKBytes * 1024
+	return nil
+}
+
+func (w *wssCollector) Destroy() {}
+
+func (w wssCollector) getPids() ([]string, error) {
+	file, err := os.Open(filepath.Join(w.cgroupCPUPath, cgroupProcfs))
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	pids := make([]string, 0)
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		pids = append(pids, scanner.Text())
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("Error in reading PIDs, err: %s", err)
+	}
+
+	if len(pids) == 0 {
+		return nil, fmt.Errorf("Not found any PID for %s cgroup", w.cgroupCPUPath)
+	}
+	return pids, nil
+}
+
+func (w wssCollector) getReferenced(pids []string) (uint64, error) {
```

Comment:
This function is stateless and does not have to be member function of `wssCollector`.